src/passes/OnceReduction.cpp

```diff
@@ -0,0 +1,414 @@
+            continue;
+          }
+
+          // This is not a call to a "once" func. However, we may have inferred
+          // that it definitely sets some "once" globals before it returns, and
+          // we can use that information.
+          for (auto globalName :
+               optInfo.onceGlobalsSetInFuncs.at(call->target)) {
+            onceGlobalsWritten.insert(globalName);
+          }
+        } else {
+          WASM_UNREACHABLE("invalid expr");
+        }
+      }
+    }
+
+    // As a result of the above optimization, we know which "once" globals are
+    // definitely written in this function. Regardless of whether this is a
+    // "once" function itself, that set of globals can be used in further
+    // optimizations, as any call to this function must set those.
+    // TODO: Aside from the entry block, we could intersect all the exit blocks.
+    optInfo.newOnceGlobalsSetInFuncs[func->name] =
+      std::move(onceGlobalsWrittenVec[0]);
+  }
+
+private:
+  OptInfo& optInfo;
+};
+
+} // anonymous namespace
+
+struct OnceReduction : public Pass {
+  void run(PassRunner* runner, Module* module) override {
+    OptInfo optInfo;
+
+    // Fill out the initial data.
+    for (auto& global : module->globals) {
+      // For a global to possibly be "once", it must be initialized to a
+      // constant. As we scan code we will turn this into false if we see
+      // anything that proves the global is not "once".
+      //   * Note that we don't check that the constant is zero - that is fine
+      //     for us to optimize, though it does indicate that the once function
+      //     will never ever run, which we could optimize further. TODO
+      //   * TODO: non-integer types?
+      optInfo.onceGlobals[global->name] = global->type.isInteger() &&
+                                          !global->imported() &&
+                                          global->init->is<Const>();
+    }
+    for (auto& func : module->functions) {
+      // Fill in the map so that it can be operated on in parallel.
+      optInfo.onceFuncs[func->name] = Name();
+    }
+
+    // Scan the module to find out which globals and functions are "once".
+    Scanner(optInfo).run(runner, module);
+
+    // Combine the information. We found which globals appear to be "once", but
+    // other information may have proven they are not so, in fact. Specifically,
+    // for a function to be "once" we need its global to also be such.
+    for (auto& kv : optInfo.onceFuncs) {
+      Name& onceGlobal = kv.second;
+      if (onceGlobal.is() && !optInfo.onceGlobals[onceGlobal]) {
+        onceGlobal = Name();
+      }
+    }
+
+    // Optimize using what we found. Keep iterating while we find things to
+    // optimize, which we estimate using a counter of the total number of once
+    // globals set by functions: as that increases, it means we are propagating
+    // useful information.
+    // TODO: limit # of iterations?
```

Comment:
It might be faster to first propagate the onceGlobalsSetInFuncs through the call graph (which doesn't require scanning through the blocks or expressions of a function) and only then start modifying functions. I think separating these concerns would make the code easier to read as well.

Reply:
Sounds good, adding a comment.